Comlex data types processing

ofekby opened the pull request:
[Resolves part of issue #121](https://github.com/dropbox/PyHive/issues/121)

This pull request aims to enable the conversion of complex types support for presto.
* Cast structs (known as ROW data type in presto) to a dictionary.
* Cast the nested varbinary values from base64 encoded strings to bytes.

Note: This pull request could have been a breaking change. To avoid it the process_complex_columns parameter was added to the cursor with default value of False.

bkyryliuk commented:
@ofekby did a quick pass on the diff, it looks really promising, thank you for contributing!

2 high level thoughts here, this class hierarchy is not very pythonic feels more java like. I seems like it could be rewritten via just using the functions and be in a smaller number of files.

Also could you please add integration tests as well e.g. run actual sql queries ?

ofekby commented:
It might be possible to minimize the number of files, although I don't think changing it would give real profit since the code is already covered with unittests and it works...

Regarding the integration tests - I would add a few in the next days

ofekby commented:
* Added an integration test with complex new table.
* Fixed some issues with the map processing (casting map keys to corresponding python type)

If you @bkyryliuk could please review or assign someone to review, thanks.

kakipipi23 commented:
Wow this PR is looking game changing for the whole Presto community!

ofekby commented:
* Removed the class inheritance
* Changed the processors to be a functions

@bkyryliuk please provide more detailed comments if there is any issue with the code

boazberman commented:
Much needed and appreciated! Thanks!

orstav commented:
Wow thats super important feature can't wait until it get merged

yana33083 commented:
Thank you so much for adding this!

ilanb1996 commented:
Waiting for this for 6 months!!!

GalDayan commented:
Looks great, thank you!

boazberman commented:
Any updates?

hashhar commented:
@bkyryliuk Is there anything the community can do to help push this over the finish line? We need this and I really want to avoid maintaining an internal fork with this patch applied.

boazberman commented:
This is super important, isn't there anything that can be done to prompt the merge request? @bkyryliuk

GalDayan commented:
@bkyryliuk Any updates?

panoptikum commented:
+1. Thanks for the contribution

bkyryliuk reviewed (approved):
PR looks good, please resolve the conflict & I'll merge and will cut new dev release

bkyryliuk commented:
@ofekby please make sure that https://github.com/ofekby/presto-types-parser is py2 friendly 
@hashhar if you can test it in your dev / prod env - it would be great, I don't have capacity to do it right now

bkyryliuk commented:
I am on parental leave now, @ashishgandhi  / @m-kogan  could someone take a look ?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dropbox/PyHive?pullRequest=328) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dropbox/PyHive?pullRequest=328) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dropbox/PyHive?pullRequest=328) it.</sub>

ofekby commented:
> I am on parental leave now, @ashishgandhi / @m-kogan could someone take a look ?

@ashishgandhi / @m-kogan  if you could please review the changes, and enable back the CI process for this PR.

ottomata commented:
What's the status here!  Waiting for this at the Wikimedia Foundation!
